### ISSUE TITLE
Wrong Igs is computed for abelian group

### ISSUE THREAD

**Issue by fingolfin** [gap-packages/polycyclic#38]
The following issue was originally reported by Sam Tertooy; I merely reduce it to a minimal working example. Consider this input
```gap
g := AbelianPcpGroup(3);
h := Subgroup(g, [ g.1, g.1^-1*g.2, g.2^2*g.3 ]);
Index(g,h);
g=h;
```
The correct answer is of course index 1 and that g and h are equal. However, with the current master branch of polycyclic, we get this instead (note that orders printed for `h` are already wrong): 
```
gap> g := AbelianPcpGroup(3);
Pcp-group with orders [ 0, 0, 0 ]
gap> h := Subgroup(g, [ g.1, g.1^-1*g.2, g.2^2*g.3 ]);
Pcp-group with orders [ 0, 0 ]
gap> Index(g,h);
infinity
gap> g=h;
false
```

According to a `git bisect`, the issue was introduced by the `AddToIgs` rewrite in 901f1a5ae52754d4e6399c4609bce98be8c3db18

**Comment by fingolfin** [comment]
Ping @beick

**Comment by sTertooy** [comment]
I went delving into the code a bit - commit 901f1a5ae52754d4e6399c4609bce98be8c3db18 was done to correct `TorsionSubgroup`. It seems the violating line of code was `g := g^0;` (line 155). Simply commenting out this line would have fixed the code, but that might have caused other errors - I don't fully understand this algorithm, so I'm not sure about the implications.

In @beick's fix, this line is moved to line 174, under a slightly different condition: `g=ind[d]`. What may have caused the new bug, however, is the following. In the original code:
```
else
    # reduce g with h
    a := LeadingExponent(g);
    b := LeadingExponent(h);
    if e.coeff1 <> 0 then
        k := (g^e.coeff1) * (h^e.coeff2);
        ind[d] := NormedPcpElement(k);
        Add(f, d);
    fi;
    # adjust g
    g := (g^e.coeff3) * (h^e.coeff4);
```
In the last step, we use `h` and `e`, the former being the original `ind[d]` and the latter depending on `h` through `b`. In the new code, the line `g := (g^e.coeff3) * (h^e.coeff4);` is essentially replaced by:
```
b := LeadingExponent(ind[d]);
e := Gcdex(a,b);
g := g^e.coeff3 * ind[d]^e.coeff4;
```
This means that, if the condition `e.coeff1 <> 0` is true, we are no longer using `h` (which was the original `ind[d]`) and the `e` depending on this `h`, but new values based on the line `ind[d] := NormedPcpElement(k);`.

After messing around a bit, I've come up with the following code.
```
while d < c do

    h := ind[d];
    r := FactorOrder(g);
    a := LeadingExponent(g);

    # shift in
    if IsBool(h) then 
        ind[d] := NormedPcpElement(g);
        Add(f,d);
        h := ind[d];
    elif not IsPrime(r) then
        b := LeadingExponent(h);
        e := Gcdex(a, b);
        if e.coeff1 <> 0 then 
            ind[d] := NormedPcpElement((g^e.coeff1)*(h^e.coeff2));
            Add(f,d);
        fi;
    fi;
			
    if g = h then 
        g := g^0;
    else
        b := LeadingExponent(h);
        e := Gcdex(a, b);
        g := g^e.coeff3 * h^e.coeff4;
    fi;
    d := Depth(g);
od;
```
It seems to fix this issue as well as #40, `TorsionSubgroup` still works correctly and it passes all tests done by `testall.g`. Then again, since I don't properly understand what the algorithm does, this might not be a good solution...

**Comment by fingolfin** [comment]
@sTertooy that looks quite plausible to me! Interested in submitting this as a PR (ideally with a test or two, based on this issue and/or issue #40)?

**Comment by fingolfin** [comment]
Fixed via #43
